### ISSUE TITLE
(possible) break in xml-enc semantics with inherited namespaces. 

### ISSUE THREAD

**Issue by chowells79** [saml-kit/xml-kit#13]
First off, I'm not 100% sure about this one. I've read the xml-enc spec, but all the places I could see this issue directly addressed were in sections explicitly called out as non-normative. Still, my best inference is that the non-normative portions were the algorithms being described - the semantics reflected should be valid.

That said, I think encrypted blocks are not handling inherited namespaces properly. I believe that if an EncryptedData element decrypts to xml content using namespaces that aren't defined within the decrypted content, the spec says to use namespaces in the current parsing context at the point the EncryptedData element was found.

This came up when generating documents that declared all their namespaces at the top level, then contained an encrypted element which contained a signature. The different data structure used to construct the encrypted element does not contain the namespace context for where it is being inserted in the parent document. This broke the signature, because the normalization algorithm treated the namespace identifiers as unknown, rather than as references to the namespaces declared in the root.

I worked around this by re-specifying the namespaces in the encrypted element. But my reading of the spec suggests that isn't strictly necessary, and the way I wanted to do it should work.

Is my reading right? If so, is this a thing that could be added without a ridiculous amount of work?

**Comment by mokhan** [comment]
I will read up the spec and follow up. Thank you for another great issue.

**Comment by mokhan** [comment]
Yes, this is definitely an issue. The xml namespaces can be defined in any ancestor element. The main cause of this defect is the use of `Hash.from_xml` here https://github.com/saml-kit/xml-kit/blob/bb655a7c7599fc5dd2fb72da0be2dd73d5e5201d/lib/xml/kit/decryption.rb#L30. This loses the possible namespace declarations in ancestor elements.

Unfortunately, I don't have a quick solution for this. I think the next steps is to update the `Decryption` class API to accept a nokogiri node element so that namespace information is not lost.

I will begin exploring this API.
